### ISSUE TITLE
Overloading function app puts app in bad state

### ISSUE THREAD

**Issue by ethanroday** [Azure/azure-functions-host#2627]
After evidently overloading a function app with ~100 requests from the same client in 5 seconds during a simulated load test, the app started returning 502 errors. Restarting did not help, and memory usage was abnormaly high. 12 hours later, the app is still not functional. All requests result in 500 errors, which appear to be caused by timeouts. The timeouts occur because a significant amount of time (generally about 40 seconds) elapses between when the function is scheduled to execute (either with `CreateNewAsync` or `RaiseEventAsync`) and when it actually executes.

#### Investigative information

- Relevant time period: 2018-04-04 00:30:00 to now
- Function App version (1.0 or 2.0-beta): 2.0-beta
- Region: West US

Specific instance for 502 error:
- Timestamp: 2018-04-04T00:28:24.059
- Invocation ID: 7b6464a4-3929-47ca-825a-b9e4741cb28a

Specific instance for 500 error:
- Timestamp: 2018-04-04T17:22:06.814
- Invocation ID: 6d9e0c25-e30b-42bc-9272-b0237d60569d

#### Known workarounds

Add the following to `host.json`:

    {
      "durableTask": {
        "disableStartInstancePolling": true
      }
    }

This mitigates the bad status, but does not fix the timeout issue.

#### Related information 

Copied from [here](https://github.com/Azure/azure-functions-durable-extension/issues/228):

Yesterday, I simluated a load test for a function app where one client sent ~100 requests in 5 seconds. The load seems to have overwhelmed the app, and I started getting 502 responses with the following (somewhat unhelpful) error message:

    The specified CGI application encountered an error and the server terminated the process.

The error has been documented in a [number](https://blogs.msdn.microsoft.com/waws/2016/12/11/asp-net-core-slow-perf-issues-on-azure-websites/) of [other](https://github.com/aspnet/KestrelHttpServer/issues/939) [places](https://stackoverflow.com/questions/33645242/the-specified-cgi-application-encountered-an-error-and-the-server-terminated-the) but, unlike in those cases, restarting the app does not make the problem go away, and I have no `web.config` to modify.

Initially, looking at Application Insights about five minutes after the errors began, there were no exceptions or failed requests being reported. However, when I looked again this morning, I can see that exceptions were indeed logged in the form of timeout errors with the following message: `Timeout expired while waiting for the new instance to start. This can happen if the task hub is overloaded or if the orchestration host failed to process the start message. Please check the orchestration logs to see whether an internal failure may have occurred.` The orchestration logs do not show any errors, though, and there are no error messages or traces in the application logs themselves (i.e. anything in the `LogFiles` directory).

Beyond debugging difficulties, my app now seems to be in a bad state. All requests are resulting in 500 errors with no messages (but with the corresponding timeout exceptions in Application Insights), even for entirely new requests that would have no checkpoint history. The function app is clearly running, but I can't make any successful requests, even in the portal. The app service for the function app also shows a large (and increasing?) memory usage for about 12 hours after the load test, despite the fact that no new requests were made in that time (approximate time of the load test is circled in red):

![image](https://user-images.githubusercontent.com/896670/38318534-65a07e8e-37e4-11e8-8aef-c70892a56c50.png)

As additional information, the data in Azure Storage Explorer seems to indicate that the orchestration functions associated with this morning's requests (the ones returning 500 errors) actually executed successfully, and output the intended results, despite Application Insights showing timeout errors for those same orchestration functions. Directly querying the functions using the HTTP API also shows that they completed successfully with the intended output.

Azure Storage Explorer:
![image](https://user-images.githubusercontent.com/896670/38321749-4dca0ec0-37ed-11e8-802e-d6ef420fe644.png)

Application Insights:
![image](https://user-images.githubusercontent.com/896670/38321885-acb850cc-37ed-11e8-887c-e26dd6a78bcb.png)

**Comment by paulbatum** [comment]
My best guess is that you hit this:
https://github.com/Azure/azure-functions-host/issues/2243

I would suggest retesting once thats fixed.

**Comment by ethanroday** [comment]
Thanks. Not sure this is the same issue, though, for a couple of reasons:
1. I never lost access to the portal. I could make calls, they just didn't complete successfully. I was also able to restart the app from the portal, although that didn't fix the issue.
2. The functions host was definitely running, at least during the period that I was getting 500 errors. Despite the long delay between scheduling and execution, the functions did execute eventually and successfully.

**Comment by paulbatum** [comment]
I've looked at the logs, I think your assessment is correct, its not related to the issues that occur when a function execution timeout is exceeded.

From the logs, I don't really see much that indicates the application was in a bad state. I do see a large number of `Timeout expired while waiting for the new instance to start.` as discussed already. I also see:

```
2018-04-04 17:20:54.1327699
MAIN_100: Function 'MainOrchestrator (Orchestrator)', version '' failed with an error. Reason: System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.) ---> 
System.NullReferenceException: Object reference not set to an instance of an object.
   at Submission#0.CreateCallbackMessage(Message incomingMessage, Object payload) in D:\home\site\wwwroot\shared\Message.csx:line 49
   at Submission#0.<SendCallbackMessage>d__5.MoveNext() in D:\home\site\wwwroot\shared\Message.csx:line 0
```
That looks like some type of application error.

You say you can't make any successful requests from the portal, but are you making any non durable requests? What if you add a simple HTTP triggered function to the app and try to invoke it?

**Comment by ethanroday** [comment]
Yes, the error you pasted above was just a bug and is unrelated 😁 

To your question, since my app has now returned to a normal state, I can't say for sure. My suspicion would be that, during the time the application was returning 502, a call to a normal function would have also returned 502, since the error I was seeing was coming from Kestrel. But during the time the application was returning 500, I'd guess that a normal function would have been able to run successfully, since the timeouts were coming from the `DurableTask` framework.

**Comment by paulbatum** [comment]
Okay well on that basis I do not think I have anything more to contribute to this. It seems like the app was in a bad state (before and after restart) from a durable functions perspective. @cgillum is there anything more I should be looking at here?
